Add additional sanity checks to check_config.h

sbutcher-arm opened the pull request:
## Description
This pull request adds additional sanity checks into `check_config.h` to ensure:
    * if test certificates are included (`MBEDTLS_CERTS_C`) there must be also be public key algorithms (`MBEDTLS_PK_C`), as the certificates will either be signed by RSA or ECDSA, and therefore need to be part of the library.
    * if any of the TLS protocols are defined (`MBEDTLS_SSL_PROTO_xxx`) then a key exchange method must also be defined (`MBEDTLS_KEY_EXCHANGE_xxx`).

Anyone who knows the library will probably not make these mistakes or will quickly diagnose and fix them, but it is possible to build combinations of both configurations (by only building the library, not the examples), and therefore users may not realise immediately that they've made a configuration mistake.

## Status
**READY**

## Requires Backporting
Yes - it's really a fix, so is worth backporting.  
Which branch?
`mbedtls-2.16` and `mbedtls-2.1`

## Migrations
NO

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

gilles-peskine-arm commented on include/mbedtls/check_config.h:
```diff
@@ -293,6 +293,10 @@
 #error "!MBEDTLS_SSL_KEEP_PEER_CERTIFICATE requires MBEDTLS_SHA512_C, MBEDTLS_SHA256_C or MBEDTLS_SHA1_C"
 #endif
 
+#if defined(MBEDTLS_CERTS_C) && !defined(MBEDTLS_PK_C)
```
Should `MBEDTLS_CERTS_C` require `MBEDTLS_X509_CRT_PARSE_C`, and `MBEDTLS_X509_CRT_PARSE_C` (and other X509 modules) require `MBEDTLS_PK_C`?

sbutcher-arm replied:
That's a fair suggestion. I've changed the dependency to now be on `MBEDTLS_X509_USE_C`, as the X.509 self-test that uses `MBEDTLS_CERTS_C` only depends on `MBEDTLS_X509_USE_C` not `MBEDTLS_X509_CRT_PARSE_C`.

gilles-peskine-arm commented on include/mbedtls/check_config.h:
```diff
@@ -586,6 +590,23 @@
+      defined(MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED) ||                      \
+      defined(MBEDTLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED) ||                    \
+      defined(MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED) )
+#error "One or more TLS protocols defined with MBEDTLS_SSL_PROTO_TLSx_x, " \
```
This line causes `check-names.sh` to think that we're using the identifier `MBEDTLS_SSL_PROTO_TLS` so it complains about a likely typo. The culprit is the line
```
sed -n 's/MBED..._[A-Z0-9_]*/\'"$NL"'&\'"$NL"/gp \
    $HEADERS library/*.c \
```

sbutcher-arm replied:
If I had time and were brave, I'd rewrite `check-names.sh`, but I'm afraid that's too bigger a task, so I just changed the error to be less specific.

Someone on the core team really should look at `check-names.sh` sometime.

gilles-peskine-arm replied:
https://github.com/ARMmbed/mbedtls/pull/1638 . Yes, we should finish that.

sbutcher-arm commented:
I've rebased and reworked. I'll backport once this is approved and passing CI.

@gilles-peskine-arm / @Patater - This is a trivial PR that I'd like to clear down - could you please indulge me with a re-review?

mpg commented:
The CI only failed the Mbed OS test with are currently not expected to work, and and instance of a reconnect test in `ssl-opt.sh`, which is a known issue with a [fix pending](https://github.com/ARMmbed/mbedtls/pull/3132), so none are blockers for merging.
